NIFI-1682 Adding RollingWindowOperation processor

JPercivall opened the pull request.

pvillard31 commented on nifi-nar-bundles/pom.xml:
```diff
@@ -55,13 +55,13 @@
         <module>nifi-elasticsearch-bundle</module>
         <module>nifi-amqp-bundle</module>
-	<module>nifi-splunk-bundle</module>
+	    <module>nifi-splunk-bundle</module>
```
really minor: could this change be reverted as it is unnecessary?

pvillard31 commented on nifi-nar-bundles/pom.xml:
```diff
@@ -55,13 +55,13 @@
         <module>nifi-spring-bundle</module>
         <module>nifi-hive-bundle</module>
-	<module>nifi-site-to-site-reporting-bundle</module>
+	    <module>nifi-site-to-site-reporting-bundle</module>
```
same comment as above

pvillard31 commented on nifi-nar-bundles/nifi-stateful-analysis-bundle/nifi-stateful-analysis-processors/src/main/java/org/apache/nifi/processors/stateful/analysis/AttributeRollingWindow.java:
```diff
@@ -0,0 +1,325 @@
+        }
+
+
+        final Double currentFlowFileValue = context.getProperty(VALUE_TO_TRACK).evaluateAttributeExpressions(flowFile).asDouble();
```
In case of NumberFormatException, I think that the flow file should be sent to a specific relationship, otherwise it will remain in the input relationship.

pvillard31 commented on nifi-nar-bundles/nifi-stateful-analysis-bundle/nifi-stateful-analysis-processors/src/main/java/org/apache/nifi/processors/stateful/analysis/AttributeRollingWindow.java:
```diff
@@ -0,0 +1,325 @@
+                Long batchCount = Long.valueOf(state.get(timeStampString+COUNT_APPEND_KEY));
+                if (!newBatch && timeStampString.equals(currBatchStart)) {
+
+                    final Double currentFlowFileValue = context.getProperty(VALUE_TO_TRACK).evaluateAttributeExpressions(flowFile).asDouble();
```
same comment regarding java.lang.NumberFormatException

pvillard31 commented on nifi-nar-bundles/nifi-stateful-analysis-bundle/nifi-stateful-analysis-processors/src/main/java/org/apache/nifi/processors/stateful/analysis/AttributeRollingWindow.java:
```diff
@@ -0,0 +1,325 @@
+
+        flowFile = session.putAllAttributes(flowFile, attributesToAdd);
+
+        session.transfer(flowFile, REL_SUCCESS);
```
send a modifyAttributes provenance event?

JPercivall replied:
In the absence of any other provenance event being emitted and attributes are modified the session will automatically emit a modifyAttribute provenance event.

pvillard31 commented on nifi-nar-bundles/nifi-stateful-analysis-bundle/nifi-stateful-analysis-processors/src/main/java/org/apache/nifi/processors/stateful/analysis/AttributeRollingWindow.java:
```diff
@@ -0,0 +1,325 @@
+    public static final int COUNT_APPEND_KEY_LENGTH = 6;
+
+    static final PropertyDescriptor VALUE_TO_TRACK = new PropertyDescriptor.Builder()
+            .name("Value to track")
```
Use .displayName() in addition to .name()?

pvillard31 commented on nifi-nar-bundles/nifi-stateful-analysis-bundle/nifi-stateful-analysis-processors/src/main/java/org/apache/nifi/processors/stateful/analysis/AttributeRollingWindow.java:
```diff
@@ -0,0 +1,325 @@
+        "with the count of FlowFiles and total aggregate value of values processed in the current time window.")
+@WritesAttributes({
+        @WritesAttribute(attribute = ROLLING_WINDOW_VALUE_KEY, description = "The rolling window value (sum of all the values stored)."),
+        @WritesAttribute(attribute = ROLLING_WINDOW_COUNT_KEY, description = "The count of the number of FlowFiles seen in the rolling window.")
```
I know this would be a really easy UpdateAttribute step to add... but could we also write another attribute containing the mean? I believe this could be useful to ensure some kind of data quality and check that some data is staying within a given range.

pvillard31 commented on nifi-nar-bundles/nifi-stateful-analysis-bundle/nifi-stateful-analysis-processors/src/test/java/org/apache/nifi/processors/stateful/analysis/TestAttributeRollingWindow.java:
```diff
@@ -0,0 +1,187 @@
+        flowFile.assertAttributeEquals("rolling_window_count", "4");
+
+    }
+
```
Add a small unit test regarding a bad format input value?

JPercivall commented:
Hey @pvillard31, thanks for the feedback. I believe I've addressed all the comments. For the exception handling I created another relationship. The set state fail relationship will typically be routed back to be retried while the "failure" would be routed off. 

I also adjusted the "get" state failures to match UpdateAttribute (it is put back on the queue). This is done because any problems with getting the state are framework related and "punishing" the FlowFile for framework problems is counterproductive.

trixpan commented:
@JPercivall - needs rebasing again. :-(

JPercivall commented:
Hey @trixpan, sorry for the delay but I just rebased it again.

pvillard31 commented:
I'm a 👍, will merge @JPercivall
